common/lib/xmodule/xmodule/contentstore/mongo.py

```diff
@@ -247,19 +247,64 @@ def _get_all_content_for_course(self,
             contentType: The mimetype string of the asset
             md5: An md5 hash of the asset content
         '''
-        query = query_for_course(course_key, "asset" if not get_thumbnails else "thumbnail")
-        find_args = {"sort": sort}
-        if maxresults > 0:
-            find_args.update({
-                "skip": start,
-                "limit": maxresults,
-            })
+        # TODO: Using an aggregate() instead of a find() here is a hack to get around the fact that Mongo 3.2 does not
+        # support sorting case-insensitively.
+        # If a sort on displayname is requested, the aggregation pipeline creates a new field:
+        # `insensitive_displayname`, a lowercase version of `displayname` that is sorted on instead.
+        # Mongo 3.4 does not require this hack. When upgraded, change this aggregation back to a find and specifiy
+        # a collation based on user's language locale instead.
+        pipeline_stages = []
+        query = query_for_course(course_key, 'asset' if not get_thumbnails else 'thumbnail')
         if filter_params:
             query.update(filter_params)
+        pipeline_stages.append({'$match': query})
+
+        if sort:
+            sort = dict(sort)
+            if 'displayname' in sort:
+                pipeline_stages.append({
+                    '$project': {
+                        'contentType': 1,
+                        'locked': 1,
+                        'chunkSize': 1,
+                        'content_son': 1,
+                        'displayname': 1,
+                        'filename': 1,
+                        'length': 1,
+                        'import_path': 1,
+                        'uploadDate': 1,
+                        'thumbnail_location': 1,
+                        'md5': 1,
+                        'insensitive_displayname': {
+                            '$toLower': '$displayname'
+                        }
+                    }
+                })
+                sort = {'insensitive_displayname': sort['displayname']}
+            pipeline_stages.append({'$sort': sort})
+
+        # This is another hack to get the total query result count, but only the Nth page of actual documents
```

Comment:
I'm not sure I understand this. I thought the API already responds with the totalCount of items for the request parameters, no? Is this something else?

Reply:
Right, the API returns a totalCount, but it needs to get the number of ALL items, not just the length of items on the page returned. The Mongo `find()` query conveniently returns this for us, but since we are using `aggregate()` now I had to calculate and add the number to the result myself.